ChangeSignature: select parameter based on cursor position

jcouv opened the pull request:
Previously, the ChangeSignature dialog would always open with the default selection (ie. the first movable or removable parameter).
With this PR, if you select a parameter in a declaration and invoke ChangeSignature, that parameter will be selected.

![image](https://user-images.githubusercontent.com/12466233/52904696-7b277a00-31e4-11e9-91bf-2141f894fa69.png)

![image](https://user-images.githubusercontent.com/12466233/52904704-89759600-31e4-11e9-849c-0a6fb077f465.png)

If this PR is ok so far, I'll make the same change on the VB side.

jcouv commented on src/EditorFeatures/CSharpTest/ChangeSignature/ReorderParametersTests.cs:
```diff
@@ -265,7 +265,8 @@ public static void M(this C goo, int y, int x, string c = ""test_c"", string b =
 
-            await TestChangeSignatureViaCommandAsync(LanguageNames.CSharp, markup, updatedSignature: permutation, expectedUpdatedInvocationDocumentCode: updatedCode);
+            await TestChangeSignatureViaCommandAsync(LanguageNames.CSharp, markup, updatedSignature: permutation,
+                expectedUpdatedInvocationDocumentCode: updatedCode, expectedSelectedIndex: 0);
```
:memo: although the parameter config has `0` for the `SelectedIndex`, the UI dialog will make an adjustment and select parameter `y` instead before the `this` parameter cannot be moved or removed.

CyrusNajmabadi commented on src/Features/CSharp/Portable/ChangeSignature/CSharpChangeSignatureService.cs:
```diff
@@ -122,13 +123,60 @@ internal sealed class CSharpChangeSignatureService : AbstractChangeSignatureServ
+                selectedIndex = GetParameterIndex(parameters.Parameters, position);
+            }
+
+            return selectedIndex;
```
nit:

```return parameters == null ? 0 : ...```

sharwell commented on src/EditorFeatures/TestUtilities/ChangeSignature/AbstractChangeSignatureTests.cs:
```diff
@@ -62,7 +62,8 @@ protected override ParseOptions GetScriptOptions()
             bool verifyNoDiagnostics = false,
-            ParseOptions parseOptions = null)
+            ParseOptions parseOptions = null,
+            int expectedSelectedIndex = -1)
```
💡 Might be better to default this to `0` so it's always covered by testing.

jcouv replied:
I tried, but in a number of tests the `ParameterConfiguration` is null. 
I don't want to make the assertion conditional on null, as that could hide issues, so I think the current design is safer.

---
In reply to: [260739080](https://github.com/dotnet/roslyn/pull/33453#discussion_r260739080) [](ancestors = 260739080)

sharwell commented on src/EditorFeatures/TestUtilities/ChangeSignature/ChangeSignatureTestState.cs:
```diff
@@ -80,6 +81,15 @@ public ChangeSignatureResult ChangeSignature()
+            WpfTestRunner.RequireWpfFact($"{nameof(AbstractChangeSignatureService.ChangeSignature)} currently needs to run on a WPF Fact because it's factored in a way that tries popping up UI in some cases.");
+
+            var context = ChangeSignatureService.GetContextAsync(InvocationDocument, _testDocument.CursorPosition.Value, restrictToDeclarations: false, CancellationToken.None)
+                .WaitAndGetResult(CancellationToken.None);
```
💡 Would be better to make the containing method asynchronous.

sharwell commented on src/Features/CSharp/Portable/ChangeSignature/CSharpChangeSignatureService.cs:
```diff
@@ -122,13 +123,19 @@ internal sealed class CSharpChangeSignatureService : AbstractChangeSignatureServ
+        private static int TryGetSelectedIndexFromDeclaration(int position, SyntaxNode matchingNode)
+        {
+            var parameters = matchingNode.ChildNodes().OfType<BaseParameterListSyntax>().SingleOrDefault();
+            return parameters == null ? 0 : GetParameterIndex(parameters.Parameters, position);
```
💡 Consider inverting this conditional to emphasize the success case.

sharwell commented on src/Features/Core/Portable/ChangeSignature/AbstractChangeSignatureService.cs:
```diff
@@ -533,5 +533,40 @@ private static SignatureChange CreateCompensatingSignatureChange(ISymbol declara
+        /// Returns 0 as the default value. Note that the ChangeSignature dialog adjusts the selection for
+        /// the `this` parameter in extension methods (the selected index won't remain 0).
+        /// </summary>
+        protected static int GetParameterIndex<TNode>(SeparatedSyntaxList<TNode> parameters, int position) where TNode : SyntaxNode
```
:bulb:

```suggestion
        protected static int GetParameterIndex<TNode>(SeparatedSyntaxList<TNode> parameters, int position)
            where TNode : SyntaxNode
```

jcouv commented:
@sharwell Thanks for the review.
FYI, I addressed all of your suggestions except for one. Left a comment to explain why that one didn't work well.
I'll go ahead and merge once CI is green.
